src/kernel/arch/x86_64/arch.zig

```diff
@@ -0,0 +1,458 @@
+///
+/// Initialise the system's memory. Populates a memory profile with boot modules from grub, the amount of available memory, the reserved regions of virtual and physical memory as well as the start and end of the kernel code
+///
+/// Arguments:
+///     IN boot_payload: BootPayload - The stivale2 info passed by limline.
+///
+/// Return: MemProfile
+///     The constructed memory profile.
+///
+/// Error: Allocator.Error
+///     Allocator.Error.OutOfMemory - There wasn't enough memory in the allocated created to populate the memory profile, consider increasing mem.FIXED_ALLOC_SIZE
+///
+pub fn initMem(boot_payload: BootPayload) Allocator.Error!MemProfile {
+    log.info("Init\n", .{});
+    defer log.info("Done\n", .{});
+
+    log.debug("KERNEL_ADDR_OFFSET:    0x{X}\n", .{@ptrToInt(&KERNEL_ADDR_OFFSET)});
+    log.debug("KERNEL_STACK_START:    0x{X}\n", .{@ptrToInt(&KERNEL_STACK_START)});
+    log.debug("KERNEL_STACK_END:      0x{X}\n", .{@ptrToInt(&KERNEL_STACK_END)});
+    log.debug("KERNEL_VADDR_START:    0x{X}\n", .{@ptrToInt(&KERNEL_VADDR_START)});
+    log.debug("KERNEL_VADDR_END:      0x{X}\n", .{@ptrToInt(&KERNEL_VADDR_END)});
+    log.debug("KERNEL_PHYSADDR_START: 0x{X}\n", .{@ptrToInt(&KERNEL_PHYSADDR_START)});
+    log.debug("KERNEL_PHYSADDR_END:   0x{X}\n", .{@ptrToInt(&KERNEL_PHYSADDR_END)});
+
+    mem.ADDR_OFFSET = @ptrToInt(&KERNEL_ADDR_OFFSET);
+
+    const allocator = &mem.fixed_buffer_allocator.allocator;
+    var reserved_physical_mem = std.ArrayList(mem.Range).init(allocator);
+    var reserved_virtual_mem = std.ArrayList(mem.Map).init(allocator);
+    // The first 640K of RAM isn't reported from the bootloader but usable
+    var memory_size_bytes: usize = 0xA0000;
+
+    // Reserve the unavailable sections from the memory map
+    // Also calculate the total RAM in the same loop
+    for (boot_payload.memmap) |entry| {
+        var entry_len = entry.len;
+        if (entry.memtype == MemType.KernelAndModules) {
+            // Kernel and modules are aligned to 4096
+            entry_len = std.mem.alignForward(entry.len, 4096);
+        }
+        if (entry.memtype == MemType.BootloaderReclaimable or entry.memtype == MemType.KernelAndModules or entry.memtype == MemType.Usable) {
+            memory_size_bytes += entry_len;
+        }
+
+        if (entry.memtype != MemType.Usable and entry.memtype != MemType.BootloaderReclaimable and entry.memtype != MemType.KernelAndModules) {
+            // If addr + len is greater than maxInt(usize) just ignore whatever comes after maxInt(usize) since it can't be addressed anyway
+            // TODO: I thing this needs to be u40 for 64 bit or u48?
+            const end: usize = if (entry.base_addr > std.math.maxInt(usize) - entry_len) std.math.maxInt(usize) else @intCast(usize, entry.base_addr + entry_len);
+            try reserved_physical_mem.append(.{
+                .start = @intCast(usize, entry.base_addr),
+                .end = end,
+            });
+        }
+    }
```

Comment:
If I understand correctly then `memory_size_bytes` will only be the amount of memory that isn't reserved. The memory profile is however expected to specify the total amount of memory, including what is reserved.

Reply:
I have reworked the memory size.